src/sparsezoo/utils/data.py

```diff
@@ -28,10 +28,158 @@
+        self._batch_size = batch_size
+        self._iterations = iterations
+        if batch_size < 0 or iterations < 0:
+            raise ValueError(
+                f"Both batch size and number of _iterations should be non-negative, "
+                f"supplied values (_batch_size, _iterations):{(batch_size, iterations)}"
+            )
+
+        self._batch_buffer = []
+        self._batch_template = self._init_batch_template()
+        self._batches_created = 0
+
+    def __iter__(self) -> Generator[List[numpy.ndarray], None, None]:
+        """
+        Iterate over the data in batches,
+        (yields from appropriate generator)
+        return: A generator for batches, each batch is enclosed in a list
+        """
+        if self._is_multi_input:
+            yield from self._multi_input_batch_generator()
+        else:
+            yield from self._single_input_batch_generator()
+
+    @property
+    def _is_multi_input(self) -> bool:
+        """
+        Check if data consists of multi-input elements
+        return: True if data consists of multi-input elements else False
+        """
+        return type(self._data[0]) is not numpy.ndarray
+
+    @property
+    def _buffer_is_full(self) -> bool:
+        """
+        Check if buffer has reached the batch size
+        """
+        return len(self._batch_buffer) == self._batch_size
+
+    @property
+    def _all_batches_loaded(self) -> bool:
+        """
+        Check if all batches have been loaded
+        """
+        return self._batches_created >= self._iterations
+
+    def _single_input_batch_generator(
+        self,
+    ) -> Generator[List[numpy.ndarray], None, None]:
+        """
+        :returns: A generator for single input batches, each element is
+            of the form [(batch_size, features)]
+        """
+        while not self._all_batches_loaded:
+            for source in self._data:
+                yield from self._batch_generator(source=source)
+                if self._all_batches_loaded:
+                    break
+
+    def _multi_input_batch_generator(
+        self,
+    ) -> Generator[List[numpy.ndarray], None, None]:
+        """
+        :returns: A generator for multi input batches, each element is
+            of the form [[(batch_size, features_a), (batch_size, features_b), ...]]
+        """
+        while not self._all_batches_loaded:
+            yield from self._batch_generator(source=self._data)
+
+    def _batch_generator(self, source) -> Generator[List[numpy.ndarray], None, None]:
+        """
+        A helper function to create batches from source
+        """
+        for sample in source:
+            self._batch_buffer.append(sample)
+            if self._buffer_is_full:
+                _batch = self._make_batch()
+                yield _batch
+                self._empty_buffer()
+                self._batches_created += 1
+                if self._all_batches_loaded:
+                    break
+
+    def _empty_buffer(self):
```

Comment:
nit: this is only used once, can we remove it and just write `self._batch_buffer = []` above?

Reply:
resolved